cmd/cmd.go

```diff
@@ -70,3 +80,338 @@ func validateUserConfig(cfg *viper.Viper) error {
+func (d *download) exercise() workspace.Exercise {
+	root := d.params.workspace
+	if d.Solution.Team.Slug != "" {
+		root = filepath.Join(root, "teams", d.Solution.Team.Slug)
+	}
+	if !d.Solution.User.IsRequester {
+		root = filepath.Join(root, "users", d.Solution.User.Handle)
+	}
+	return workspace.Exercise{
+		Root:  root,
+		Track: d.Solution.Exercise.Track.ID,
+		Slug:  d.Solution.Exercise.ID,
+	}
+}
+
+func (d *download) metadata() workspace.ExerciseMetadata {
+	return workspace.ExerciseMetadata{
+		AutoApprove: d.Solution.Exercise.AutoApprove,
+		Track:       d.Solution.Exercise.Track.ID,
+		Team:        d.Solution.Team.Slug,
+		Exercise:    d.Solution.Exercise.ID,
+		ID:          d.Solution.ID,
+		URL:         d.Solution.URL,
+		Handle:      d.Solution.User.Handle,
+		IsRequester: d.Solution.User.IsRequester,
+	}
+}
+
+// validate validates the presence of an ID and checks for any error responses.
+func (d *download) validate() error {
+	if d.Solution.ID == "" {
+		return errors.New("download missing an ID")
+	}
+	if d.Error.Message != "" {
+		return errors.New(d.Error.Message)
+	}
+	return nil
+}
+
+// downloadWriter writes download contents to the workspace.
+type downloadWriter struct {
+	*download
+}
+
+// writeMetadata writes the exercise metadata.
+func (d downloadWriter) writeMetadata() error {
+	metadata := d.metadata()
+	return metadata.Write(d.destination())
+}
+
+// writeSolutionFiles attempts to write each exercise file that is part of the downloaded Solution.
+// An HTTP request is made using each filename and failed responses are swallowed.
+// All successful file responses are written except when 0 Content-Length.
+func (d downloadWriter) writeSolutionFiles() error {
+	if d.params.fromExercise {
+		return errors.New("existing exercise files should not be overwritten")
+	}
+	for _, filename := range d.Solution.Files {
+		res, err := d.requestFile(filename)
+		if err != nil {
+			return err
+		}
+		if res == nil {
+			continue
+		}
+		defer res.Body.Close()
+
+		// TODO: if there's a collision, interactively resolve (show diff, ask if overwrite).
+		// TODO: handle --force flag to overwrite without asking.
+
+		sanitizedPath := sanitizeLegacyNumericSuffixFilepath(filename, d.exercise().Slug)
+		fileWritePath := filepath.Join(d.destination(), sanitizedPath)
+		if err = os.MkdirAll(filepath.Dir(fileWritePath), os.FileMode(0755)); err != nil {
+			return err
+		}
+
+		f, err := os.Create(fileWritePath)
+		if err != nil {
+			return err
+		}
+		defer f.Close()
+		if _, err := io.Copy(f, res.Body); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// destination is the download destination path.
+func (d downloadWriter) destination() string {
+	return d.exercise().MetadataDir()
+}
+
+// downloadParams is required to create a download.
+type downloadParams struct {
+	usrCfg *viper.Viper
+	uuid   string
+	slug   string
+	track  string
+	team   string
+
+	// config
+	token, apibaseurl, workspace string
```

Comment:
I'm not sure it's worth putting the config values as fields here. The reasoning is that rather than have repeated lookups for config values, e.g. `d.usrCfg.GetString("token")`, we can store them in `downloadParams` so that `Download` has the values it needs directly without depending on the config structure. Encapsulating these values in fields has the benefit of preventing errors that the compiler can't catch, misspelling a key, say, `d.usrCfg.GetString("tokn")`.

If this seems worthwhile, `downloadParams` shouldn't need to store *viper.Viper as a field. Currently it remains because it's needed in `validate()` but that can change depending on if we want to move validation to the config object as mentioned in a further comment below. If we don't want to store it on the config object, we can add the validation logic directly to `downloadParams`.